docs/proposals/file-sync-location.md

```diff
@@ -0,0 +1,26 @@
+
+# File Sync Location for Devfile Components in Odo
+
+## Background 
+
+Currently, odo uses the following logic to determine where to sync source code to a devfile component:
+
+1) If the `sourceMapping` field is set, odo will mount the source volume under to the specified folder and sync the source code there
+
+2) If `sourceMapping` is unset, odo will do one of two things:
+   a) If a single project is defined in the devfile, odo will sync to `/projects/<projectName>
+   b) If no projects are defined, or there are multiple projects defined, odo will sync to `/projects`
+
+The downside to this current approach is that the expected sync directory isn't consistent for somebody writing a devfile, and as they add or remove projects from the devfile, the sync directory will change, potentially breaking the commands in their devfile each time.
+
+## Implementation plan
+
+To make the syncing directory more consistent, odo will do the following (in the order given):
+
+1) If the `sourceMapping` field is set, odo will mount the source volume under to the specified folder and sync the source code there. The `PROJECTS_ROOT` environment variable in each container will be set to this value (as-is today).
+
+2) If the `PROJECTS_ROOT` environment variable is set on the user's machine, odo will sync to there, *and* set it as an environment variable in each remore container
```

Comment:
It checks if the environment variable is set on user's development system? Wouldn't it ease things if we provide an interface to set this through `odo env` command instead?

Reply:
We could certainly provide that as an option, but according to the devfile spec (under `sourceMapping`), it's expected that `PROJECTS_ROOT` can be overridden via an environment variable

Reply:
This looks a little bit weird to me. Are we really want to map local env variables to the container env variables?

The way I was imagining this was that the `PROJECT_ROOT` env variable needs to be set in devfile.yaml in the container section. The benefit of using this over `soruceMapping` is that can have scripts inside container that are using this variable.

Reply:
Re-reading:
```
 Optional specification of the path in the container where project sources should be transferred/mounted when mountSources is true. When omitted, the value of the PROJECTS_ROOT environment variable is used.
```

I suppose it is referring to the value of `PROJECTS_ROOT` in the container where the source is being mounted, in which case, it could be modified by setting it as an env var in the container.

I can update the doc accordingly.

Reply:
@kadel Updated the doc accordingly.